rust/kernel/error.rs

```diff
@@ -104,3 +104,62 @@ impl From<AllocError> for Error {
         Error::ENOMEM
     }
 }
+
+// # Invariant: `-bindings::MAX_ERRNO` fits in an `i16`.
+crate::static_assert!(bindings::MAX_ERRNO as i64 <= -(i16::MIN as i64));
+
+/// Transform a kernel "error pointer" to a normal pointer.
+///
+/// Some kernel C API functions return an "error pointer" which optionally
+/// embeds an `errno`. Callers are supposed to check the returned pointer
+/// for errors. This function performs the check and converts the "error pointer"
+/// to a normal pointer in an idiomatic fashion.
+///
+/// # Safety
+///
+/// `ptr` must be valid or null.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// fn devm_platform_ioremap_resource(
+///     pdev: &mut PlatformDevice,
+///     index: u32,
+/// ) -> Result<*mut c_types::c_void> {
+///     // SAFETY: FFI call.
+///     unsafe {
+///         from_kernel_err_ptr(bindings::devm_platform_ioremap_resource(
+///             pdev.to_ptr(),
+///             index,
+///         ))
+///     }
+/// }
+/// ```
+// TODO: remove `dead_code` marker once an in-kernel client is available.
+#[allow(dead_code)]
+pub(crate) fn from_kernel_err_ptr<T>(ptr: *mut T) -> Result<*mut T> {
+    extern "C" {
+        #[allow(improper_ctypes)]
+        fn rust_helper_is_err(ptr: *const c_types::c_void) -> bool;
+
+        #[allow(improper_ctypes)]
+        fn rust_helper_ptr_err(ptr: *const c_types::c_void) -> c_types::c_long;
+    }
+
+    // SAFETY: any pointer can be safely cast to a `*const c_types::c_void`.
+    let const_ptr = ptr as *const c_types::c_void;
+    // SAFETY: `IS_ERR()` requires a valid-or-null pointer, and `const_ptr`
```

Comment:
Do we have the definition for `a valid-or-null pointer`? I think here we can just say, `IS_ERR()` accepts any pointer as the argument.

Reply:
I checked the implementation of `IS_ERR()` and `PTR_ERR()` in the kernel, and you are right, the pointer doesn't need to point to a valid memory location. Those functions do pointer arithmetic only, and never touch the memory pointed to.

Looks like `// SAFETY: FFI call` is probably all we need. And we can then eliminate the `# Safety` constraint on the function too.